### PR TITLE
Fix compute nightly version Scala 2

### LINKED ISSUE #867: New `.nightly` support retrieves wrong (too-old) versions
```
% scala-cli repl -S 2.12.nightly
Welcome to Scala 2.12.16-bin-586302a (OpenJDK 64-Bit Server VM, Java 17.0.2).
Type in expressions for evaluation. Or try :help.
```

but scala/scala@586302a is from November 2021

similar for 2.13:

```
% scala-cli repl -S 2.13.nightly
Welcome to Scala 2.13.9-bin-9312709 (OpenJDK 64-Bit Server VM, Java 17.0.2).
```

scala/scala@9312709 is a couple months old (it's from February 2022)

finally Scala 3:

```
% scala-cli repl -S 3.nightly
Welcome to Scala 3.2.0-RC1-bin-20220308-29073f1-NIGHTLY-git-29073f1 (17.0.2, Java OpenJDK 64-Bit Server VM).
```

20220308 was a month ago.

fyi @zmerr

### PR DESCRIPTION
**Pull request by lwronski**

### CODE REVIEW

File: modules/build/src/test/scala/scala/build/tests/BuildOptionsTests.scala [hunk 1]
```diff
@@ -228,7 +228,7 @@ class BuildOptionsTests extends munit.FunSuite {
     )
     val scalaParams = options.scalaParams.orThrow
     assert(
-      scalaParams.scalaVersion == "2.12.16-bin-586302a",
+      scalaParams.scalaVersion.contains("2.12"),
```
Comment (alexarchambault) [inline modules/build/src/test/scala/scala/build/tests/BuildOptionsTests.scala]: Wouldn't `scala2NightlyRegex.unapplySeq(scalaParams.scalaVersion).isDefined` like above work?

File: modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    final case class ScalaVersion(name: String, lastModified: Long)
+    final case class ScalaVersionsMetaData(repo: String, children: List[ScalaVersion])
+
+    val codec: JsonValueCodec[ScalaVersionsMetaData] = JsonCodecMaker.make
+  }
+
+  private def downloadScala2RepoPage(cache: FileCache[Task]): Either[BuildException, Array[Byte]] =
+    either {
+      val scala2NightlyRepo =
+        s"https://scala-ci.typesafe.com/ui/api/v1/ui/nativeBrowser/scala-integration/org/scala-lang/scala-compiler"
```
Comment (alexarchambault) [inline modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: NIT No need of a string interpolator here:
```suggestion
        "https://scala-ci.typesafe.com/ui/api/v1/ui/nativeBrowser/scala-integration/org/scala-lang/scala-compiler"
```

File: modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+import scala.build.EitherCps.{either, value}
+import scala.build.Os
+import scala.build.errors.{BuildException, ScalaVersionError}
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+
+object ScalaNightlyVersion {
+
+  object Scala2Repo {
```
Comment (alexarchambault) [inline modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: Couldn't this be private?

File: modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      val scala2NightlyRepo =
+        "https://scala-ci.typesafe.com/ui/api/v1/ui/nativeBrowser/scala-integration/org/scala-lang/scala-compiler"
+      val artifact = Artifact(scala2NightlyRepo).withChanging(true)
+      val res = cache.logger.use {
+        try cache.withTtl(0.seconds).file(artifact).run.unsafeRun()(cache.ec)
+        catch {
+          case NonFatal(e) => throw new Exception(e)
+        }
+      }.left.map { err =>
+        Util.printException(err)
```
Comment (alexarchambault) [inline modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: I don't think we should manually print that from there. Better pass `err` as the cause exception to `ScalaVersionError`.

File: modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    val codec: JsonValueCodec[ScalaVersionsMetaData] = JsonCodecMaker.make
+  }
+
+  private def downloadScala2RepoPage(cache: FileCache[Task]): Either[BuildException, Array[Byte]] =
+    either {
+      val scala2NightlyRepo =
+        "https://scala-ci.typesafe.com/ui/api/v1/ui/nativeBrowser/scala-integration/org/scala-lang/scala-compiler"
+      val artifact = Artifact(scala2NightlyRepo).withChanging(true)
+      val res = cache.logger.use {
+        try cache.withTtl(0.seconds).file(artifact).run.unsafeRun()(cache.ec)
```
Comment (romanowski) [inline modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: Would ttl set to 0 cause every run to access the file?

File: modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        catch {
+          case NonFatal(e) => throw new Exception(e)
+        }
+      }.left.map { err =>
+        val msg =
+          s"""|Unable to compute the latest Scala 2 nightly version.
+              |Throws error during downloading web page repository for Scala 2.
+              |   Caused by: $err""".stripMargin
+        new ScalaVersionError(msg)
```
Comment (romanowski) [inline modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: Hmm, I wonder if shouldn't be a warning and latest local matching scala version should be use.

Generally, now `--S 2.nightly` is unusable without internet access

Comment (lwronski) [reply modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: At the first time it is required internet access to download:
```
https://scala-ci.typesafe.com/ui/api/v1/ui/nativeBrowser/scala-integration/org/scala-lang/scala-compiler
```

but, if there is no access to the internet the second time, `coursier` uses previously download web page repository from cache. So, Scala CLI uses the latest local matching scala nightly version.

File: modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala (hunk 5 above)
Comment (lwronski) [reply modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: It means, the courier try to download web page repository every time. If there is no access to internet, it uses previously downloaded page.

Comment (romanowski) [reply modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: Can we increase ttl to an hour or two? So noop compilation with .nightly are fast?

**Review by Gedochao — APPROVED** [review]

File: modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala (hunk 4 above)
Comment (alexarchambault) [reply modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: I mean, passing `err` not via its message, but as the exception cause. `BuildException` accepts a `cause: Throwable` in its constructor. `ScalaVersionError` should accept it too, and we can pass the cause from here.

Comment (lwronski) [reply modules/options/src/main/scala/scala/build/internal/ScalaNightlyVersion.scala]: I missed that buildException accepts a Throwable parameter, thanks

**Review by alexarchambault — APPROVED** [review]

**Pull request merged**